File: problem/Median_Two_Sorted_Arrays/findMedianSortedArrays.cpp
class Solution
{
public:
    double findMedianSortedArrays(vector<int> &nums1, vector<int> &nums2)
    {
        if (nums1.size() > nums2.size())
        {
            return findMedianSortedArrays(nums2, nums1);
        }

        int m = nums1.size(), n = nums2.size();
        int left = 0, right = m;

        while (left <= right)
        {
            int partition1 = (left + right) / 2;
            int partition2 = (m + n + 1) / 2 - partition1;

            int maxLeft1 = (partition1 == 0) ? INT_MIN : nums1[partition1 - 1];
            int minRight1 = (partition1 == m) ? INT_MAX : nums1[partition1];

            int maxLeft2 = (partition2 == 0) ? INT_MIN : nums2[partition2 - 1];
            int minRight2 = (partition2 == n) ? INT_MAX : nums2[partition2];

            if (maxLeft1 <= minRight2 && maxLeft2 <= minRight1)
            {
                if ((m + n) % 2 == 0)
                {
                    return (max(maxLeft1, maxLeft2) + min(minRight1, minRight2)) / 2.0;
                }
                else
                {
                    return max(maxLeft1, maxLeft2);
                }
            }
            else if (maxLeft1 > minRight2)
            {
                right = partition1 - 1;
            }
            else
            {
                left = partition1 + 1;
            }
        }

        throw invalid_argument("Input arrays are not sorted");
    }

    double findMedianSortedArrays_merge(vector<int> &nums1, vector<int> &nums2)
    {
        vector<int> merged;
        int i = 0, j = 0;

        while (i < nums1.size() && j < nums2.size())
        {
            if (nums1[i] <= nums2[j])
            {
                merged.push_back(nums1[i++]);
            }
            else
            {
                merged.push_back(nums2[j++]);
            }
        }

        while (i < nums1.size())
            merged.push_back(nums1[i++]);
        while (j < nums2.size())
            merged.push_back(nums2[j++]);

        int n = merged.size();
        if (n % 2 == 0)
        {
            return (merged[n / 2 - 1] + merged[n / 2]) / 2.0;
        }
        else
        {
            return merged[n / 2];
        }
    }

    double findMedianSortedArrays_kth(vector<int> &nums1, vector<int> &nums2)
    {
        int total = nums1.size() + nums2.size();
        if (total % 2 == 1)
        {
            return findKthElement(nums1, nums2, total / 2 + 1);
        }
        else
        {
            return (findKthElement(nums1, nums2, total / 2) +
                    findKthElement(nums1, nums2, total / 2 + 1)) /
                   2.0;
        }
    }

private:
    double findKthElement(vector<int> &nums1, vector<int> &nums2, int k)
    {
        if (nums1.size() > nums2.size())
        {
            return findKthElement(nums2, nums1, k);
        }

        int m = nums1.size(), n = nums2.size();
        int left = max(0, k - n), right = min(k, m);

        while (left < right)
        {
            int cut1 = (left + right) / 2;
            int cut2 = k - cut1;

            int left1 = (cut1 == 0) ? INT_MIN : nums1[cut1 - 1];
            int left2 = (cut2 == 0) ? INT_MIN : nums2[cut2 - 1];

            int right1 = (cut1 == m) ? INT_MAX : nums1[cut1];
            int right2 = (cut2 == n) ? INT_MAX : nums2[cut2];

            if (left1 <= right2 && left2 <= right1)
            {
                return max(left1, left2);
            }
            else if (left1 > right2)
            {
                right = cut1 - 1;
            }
            else
            {
                left = cut1 + 1;
            }
        }

        int cut1 = left;
        int cut2 = k - cut1;

        int left1 = (cut1 == 0) ? INT_MIN : nums1[cut1 - 1];
        int left2 = (cut2 == 0) ? INT_MIN : nums2[cut2 - 1];

        return max(left1, left2);
    }
};

File: problem/Median_Two_Sorted_Arrays/findMedianSortedArrays.java
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        if (nums1.length > nums2.length) {
            return findMedianSortedArrays(nums2, nums1);
        }

        int m = nums1.length, n = nums2.length;
        int left = 0, right = m;

        while (left <= right) {
            int partition1 = (left + right) / 2;
            int partition2 = (m + n + 1) / 2 - partition1;

            int maxLeft1 = (partition1 == 0) ? Integer.MIN_VALUE : nums1[partition1 - 1];
            int minRight1 = (partition1 == m) ? Integer.MAX_VALUE : nums1[partition1];

            int maxLeft2 = (partition2 == 0) ? Integer.MIN_VALUE : nums2[partition2 - 1];
            int minRight2 = (partition2 == n) ? Integer.MAX_VALUE : nums2[partition2];

            if (maxLeft1 <= minRight2 && maxLeft2 <= minRight1) {
                if ((m + n) % 2 == 0) {
                    return (Math.max(maxLeft1, maxLeft2) + Math.min(minRight1, minRight2)) / 2.0;
                } else {
                    return Math.max(maxLeft1, maxLeft2);
                }
            } else if (maxLeft1 > minRight2) {
                right = partition1 - 1;
            } else {
                left = partition1 + 1;
            }
        }

        throw new IllegalArgumentException("Input arrays are not sorted");
    }

    public double findMedianSortedArrays_merge(int[] nums1, int[] nums2) {
        int[] merged = new int[nums1.length + nums2.length];
        int i = 0, j = 0, k = 0;

        while (i < nums1.length && j < nums2.length) {
            if (nums1[i] <= nums2[j]) {
                merged[k++] = nums1[i++];
            } else {
                merged[k++] = nums2[j++];
            }
        }

        while (i < nums1.length)
            merged[k++] = nums1[i++];
        while (j < nums2.length)
            merged[k++] = nums2[j++];

        int n = merged.length;
        if (n % 2 == 0) {
            return (merged[n / 2 - 1] + merged[n / 2]) / 2.0;
        } else {
            return merged[n / 2];
        }
    }

    public double findMedianSortedArrays_recursive(int[] nums1, int[] nums2) {
        int total = nums1.length + nums2.length;
        if (total % 2 == 1) {
            return findKthElement(nums1, 0, nums2, 0, total / 2 + 1);
        } else {
            return (findKthElement(nums1, 0, nums2, 0, total / 2) +
                    findKthElement(nums1, 0, nums2, 0, total / 2 + 1)) / 2.0;
        }
    }

    private double findKthElement(int[] nums1, int start1, int[] nums2, int start2, int k) {
        if (start1 >= nums1.length) {
            return nums2[start2 + k - 1];
        }
        if (start2 >= nums2.length) {
            return nums1[start1 + k - 1];
        }
        if (k == 1) {
            return Math.min(nums1[start1], nums2[start2]);
        }

        int mid1 = start1 + k / 2 - 1;
        int mid2 = start2 + k / 2 - 1;

        int val1 = (mid1 >= nums1.length) ? Integer.MAX_VALUE : nums1[mid1];
        int val2 = (mid2 >= nums2.length) ? Integer.MAX_VALUE : nums2[mid2];

        if (val1 < val2) {
            return findKthElement(nums1, mid1 + 1, nums2, start2, k - k / 2);
        } else {
            return findKthElement(nums1, start1, nums2, mid2 + 1, k - k / 2);
        }
    }
}
